# Lab #3: Democrats 2018 Primary Results and Data Analysis Updated with Pandas

Lillian Ellis

Created September 25, 2018 // Data set Revisited: October 16, 2018

**Abstract:**

Original Abstract:

>In this lab, I worked with a data set surrounding the democrats involved in the primary elections this year around our country. Throughout the lab, I worked to answer a few questions. How many candidates were running for governor? What about the U.S. House of Representatives? How many identified as LGBTQ, non-white, or were veterans? How many were Obama Alum? How many of these democratic candidates advanced to the general election? How many lost their races?

Additionally:

Additionally, this time around, I hope to gather a wider range of data and especially learn how to count unique values, so I am able to track how many candidates ran in a specific state, office type, or district. I hoped to clean my data (especially replace the null values), and be able to work through it in an overall more efficient manor as my code in Lab #2 was pretty round-about and messy. 

**Data Set Preparation:**

My data is all from FiveThirtyEight: a well known, reputable website founded by Nate Silver that focuses on data analysis, specifically poll analysis. The data set contains information from all open democratic primary elections through August 7th of 2018. The data set contains information on the 811 democratic candidates who ran for various positions up until the 7th. The specific data set I used contained 812 lines of data, and was extremely well organized. One downside of the data-set was that some information surrounding party support, Emily endorsed, gun sense, Biden endorsed, and more were missing or partially filled out. The table was not fully complete. However, the data was well organized, accurate, and the GitHub description contained information on where each number, fact, or name was pulled from.

Link to Data Set (From FiveThirtyEight's GitHub): https://github.com/fivethirtyeight/data/tree/master/primary-candidates-2018

Link to Five Thirty Eight Article using Data: https://fivethirtyeight.com/features/the-establishment-is-beating-the-progressive-wing-in-democratic-primaries-so-far/

First, I imported all the pandas packages and pulled them in as abriviated letters. In one line, I read my filed, and defined the header as the first two rows (as these are headers, not data)

Before cleaning my data with code, I created a clean file in excel as my datas headers were being read in incorrectly. I just deleted the title header before importing it in, which made my life a lot easier (as the data wasn't shifted anymore). 

In [148]:
#Importing all pandas packages 
#this way to avoid name space conflicts 
import math as m
import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns


#one line to open and read a csv filed into a data frame data structure
df3 = pd.read_csv('dem_candidates_clean.csv')

#print the first few lines of the data frame, default is 5 lines
df3.head()


candidate state             district office_type race_type  \
0  Anthony White (Alabama)    AL  Governor of Alabama    Governor   Regular   
1   Christopher Countryman    AL  Governor of Alabama    Governor   Regular   
2    Doug "New Blue" Smith    AL  Governor of Alabama    Governor   Regular   
3          James C. Fields    AL  Governor of Alabama    Governor   Regular   
4            Sue Bell Cobb    AL  Governor of Alabama    Governor   Regular   

  race_primary_election_date primary_status primary_runoff_status  \
0                     6/5/18           Lost                  None   
1                     6/5/18           Lost                  None   
2                     6/5/18           Lost                  None   
3                     6/5/18           Lost                  None   
4                     6/5/18           Lost                  None   

  general_status  partisan_lean        ...          Biden Endorsed?  \
0           None     -28.879999        ...                      NaN   
1           None     -28.879999        ...                      NaN   
2           None     -28.879999        ...                      NaN   
3           None     -28.879999        ...                      NaN   
4           None     -28.879999        ...                      NaN   

  Warren Endorsed?  Sanders Endorsed? Our Revolution Endorsed?  \
0               NaN               NaN                      NaN   
1               NaN               NaN                      NaN   
2               NaN               NaN                      NaN   
3               NaN               NaN                      NaN   
4               NaN               NaN                      NaN   

  Justice Dems Endorsed? PCCC Endorsed? Indivisible Endorsed? WFP Endorsed?  \
0                    NaN            NaN                   NaN           NaN   
1                    NaN            NaN                   NaN           NaN   
2                    NaN            NaN                   NaN           NaN   
3                    NaN            NaN                   NaN           NaN   
4                    NaN            NaN                   NaN           NaN   

  VoteVets Endorsed? No Labels Support?  
0                NaN                NaN  
1                NaN                NaN  
2                NaN                NaN  
3                NaN                NaN  
4                NaN                NaN  

[5 rows x 32 columns]

After completing lab 4, I now look at my data set and I realize that there are alot of null values which can easily be replaced. 
My next step is to replace these null values with x. 


In [158]:
#filling null values with x through the .fillna method 
df3 = df3.fillna('x')
#print head with "head" method to display replaced null values 
df3.head()


candidate state             district office_type race_type  \
0  Anthony White (Alabama)    AL  Governor of Alabama    Governor   Regular   
1   Christopher Countryman    AL  Governor of Alabama    Governor   Regular   
2    Doug "New Blue" Smith    AL  Governor of Alabama    Governor   Regular   
3          James C. Fields    AL  Governor of Alabama    Governor   Regular   
4            Sue Bell Cobb    AL  Governor of Alabama    Governor   Regular   

  race_primary_election_date primary_status primary_runoff_status  \
0                     6/5/18           Lost                  None   
1                     6/5/18           Lost                  None   
2                     6/5/18           Lost                  None   
3                     6/5/18           Lost                  None   
4                     6/5/18           Lost                  None   

  general_status  partisan_lean        ...         Biden Endorsed?  \
0           None     -28.879999        ...                       x   
1           None     -28.879999        ...                       x   
2           None     -28.879999        ...                       x   
3           None     -28.879999        ...                       x   
4           None     -28.879999        ...                       x   

  Warren Endorsed?  Sanders Endorsed? Our Revolution Endorsed?  \
0                 x                 x                        x   
1                 x                 x                        x   
2                 x                 x                        x   
3                 x                 x                        x   
4                 x                 x                        x   

  Justice Dems Endorsed? PCCC Endorsed? Indivisible Endorsed? WFP Endorsed?  \
0                      x              x                     x             x   
1                      x              x                     x             x   
2                      x              x                     x             x   
3                      x              x                     x             x   
4                      x              x                     x             x   

  VoteVets Endorsed? No Labels Support?  
0                  x                  x  
1                  x                  x  
2                  x                  x  
3                  x                  x  
4                  x                  x  

[5 rows x 32 columns]

Other than replacing those null values, there wasn't much else that needed to be done to clean this data set, as it was pretty well organized. One thing I did do was rename some of the column headers that ended in questions. I simplified these titles to a more consise and snake-case description.

Renaming headers with long names to simplify analysis:

In [150]:
#fixing the headers, some headers have long names

#with the rename method, I called the header name and renamed it

#inplace directly replaces it in your list 
#this makes a permanent change, a view (not just editing the view)

df3.rename(columns = {'LGBTQ?':'lgbtq',
                     'Race' : 'race',
                      'Veteran?' : 'veteran',
                      'Obama Alum' : 'obama'},

              inplace = True)
                      
df3.head()

candidate state             district office_type race_type  \
0  Anthony White (Alabama)    AL  Governor of Alabama    Governor   Regular   
1   Christopher Countryman    AL  Governor of Alabama    Governor   Regular   
2    Doug "New Blue" Smith    AL  Governor of Alabama    Governor   Regular   
3          James C. Fields    AL  Governor of Alabama    Governor   Regular   
4            Sue Bell Cobb    AL  Governor of Alabama    Governor   Regular   

  race_primary_election_date primary_status primary_runoff_status  \
0                     6/5/18           Lost                  None   
1                     6/5/18           Lost                  None   
2                     6/5/18           Lost                  None   
3                     6/5/18           Lost                  None   
4                     6/5/18           Lost                  None   

  general_status  partisan_lean        ...         Biden Endorsed?  \
0           None     -28.879999        ...                       x   
1           None     -28.879999        ...                       x   
2           None     -28.879999        ...                       x   
3           None     -28.879999        ...                       x   
4           None     -28.879999        ...                       x   

  Warren Endorsed?  Sanders Endorsed? Our Revolution Endorsed?  \
0                 x                 x                        x   
1                 x                 x                        x   
2                 x                 x                        x   
3                 x                 x                        x   
4                 x                 x                        x   

  Justice Dems Endorsed? PCCC Endorsed? Indivisible Endorsed? WFP Endorsed?  \
0                      x              x                     x             x   
1                      x              x                     x             x   
2                      x              x                     x             x   
3                      x              x                     x             x   
4                      x              x                     x             x   

  VoteVets Endorsed? No Labels Support?  
0                  x                  x  
1                  x                  x  
2                  x                  x  
3                  x                  x  
4                  x                  x  

[5 rows x 32 columns]

In [151]:
#test to see if it worked (can't see the renamed column headers with the head function)
#so, I called the index of the column through the list function to see if it worked, it did!
print(list(df3)[14])  

lgbtq


**Data Modeling:**

I primarily relied on the .count function throughout finding my information, and this served as extremley helpful. In the future, I would love to convert specific parts of my data to a different type in order to do even more analysis. 

**Office Type:**

In [152]:
print("Candidate Information by Office Type They Campaigned For:")

#groupby goes through all of the elements of the listed column as a group 

#the count method automatically counts the amount of every value in the column for the group I have named. 
#So the code below will count the occurance of the different office types in each column!
#I have easy access to super easy, printable information through the groupby and count methods
df3.groupby("office_type").count()



Candidate Information by Office Type They Campaigned For:


candidate  state  district  race_type  \
office_type                                             
Governor              100    100       100        100   
Representative        687    687       687        687   
Senator                24     24        24         24   

                race_primary_election_date  primary_status  \
office_type                                                  
Governor                               100             100   
Representative                         687             687   
Senator                                 24              24   

                primary_runoff_status  general_status  partisan_lean  \
office_type                                                            
Governor                          100             100            100   
Representative                    687             687            687   
Senator                            24              24             24   

                primary_pct         ...          Biden Endorsed?  \
office_type                         ...                            
Governor                100         ...                      100   
Representative          687         ...                      687   
Senator                  24         ...                       24   

                Warren Endorsed?   Sanders Endorsed?  \
office_type                                            
Governor                      100                100   
Representative                687                687   
Senator                        24                 24   

                Our Revolution Endorsed?  Justice Dems Endorsed?  \
office_type                                                        
Governor                             100                     100   
Representative                       687                     687   
Senator                               24                      24   

                PCCC Endorsed?  Indivisible Endorsed?  WFP Endorsed?  \
office_type                                                            
Governor                   100                    100            100   
Representative             687                    687            687   
Senator                     24                     24             24   

                VoteVets Endorsed?  No Labels Support?  
office_type                                             
Governor                       100                 100  
Representative                 687                 687  
Senator                         24                  24  

[3 rows x 31 columns]

Now, I can use this same count method with groupby to find a crazy amount of information for each candidate. Here I am finding the information of how many candidates advanced and how many lost. 

**Wins and Losses Count:**

In [153]:
#win and losses count:
#I can also use the same count method to see how many candidates advanced and how many lost. 
df3.groupby("primary_status").count()

candidate  state  district  office_type  race_type  \
primary_status                                                       
Advanced              265    265       265          265        265   
Lost                  546    546       546          546        546   

                race_primary_election_date  primary_runoff_status  \
primary_status                                                      
Advanced                               265                    265   
Lost                                   546                    546   

                general_status  partisan_lean  primary_pct  \
primary_status                                               
Advanced                   265            265          265   
Lost                       546            546          546   

                       ...          Biden Endorsed?  Warren Endorsed?   \
primary_status         ...                                               
Advanced               ...                      265                265   
Lost                   ...                      546                546   

                Sanders Endorsed?  Our Revolution Endorsed?  \
primary_status                                                
Advanced                      265                       265   
Lost                          546                       546   

                Justice Dems Endorsed?  PCCC Endorsed?  Indivisible Endorsed?  \
primary_status                                                                  
Advanced                           265             265                    265   
Lost                               546             546                    546   

                WFP Endorsed?  VoteVets Endorsed?  No Labels Support?  
primary_status                                                         
Advanced                  265                 265                 265  
Lost                      546                 546                 546  

[2 rows x 31 columns]

What I can do next is even more exciting: I can count how many candidates ran in each state! And what those corresponding states were. The count function counts unique values, so I've completed my goal with the following line of code. 

**Counting Unique States:**

In [154]:
#states 
df3.groupby("state").count()

candidate  district  office_type  race_type  \
state                                                
AL            16        16           16         16   
AR             9         9            9          9   
AZ             2         2            2          2   
CA            68        68           68         68   
CO            14        14           14         14   
GA            26        26           26         26   
IA            16        16           16         16   
ID             8         8            8          8   
IL            39        39           39         39   
IN            30        30           30         30   
KS            15        15           15         15   
KY            17        17           17         17   
MD            22        22           22         22   
ME            11        11           11         11   
MI            37        37           37         37   
MO            16        16           16         16   
MS            10        10           10         10   
MT             6         6            6          6   
NC            20        20           20         20   
ND             1         1            1          1   
NE            12        12           12         12   
NJ            15        15           15         15   
NM            11        11           11         11   
NV            31        31           31         31   
NY            38        38           38         38   
OH            47        47           47         47   
OK            23        23           23         23   
OR             7         7            7          7   
PA            43        43           43         43   
SC            23        23           23         23   
SD             2         2            2          2   
TN            22        22           22         22   
TX           109       109          109        109   
UT             6         6            6          6   
VA            20        20           20         20   
WA            10        10           10         10   
WV             9         9            9          9   

       race_primary_election_date  primary_status  primary_runoff_status  \
state                                                                      
AL                             16              16                     16   
AR                              9               9                      9   
AZ                              2               2                      2   
CA                             68              68                     68   
CO                             14              14                     14   
GA                             26              26                     26   
IA                             16              16                     16   
ID                              8               8                      8   
IL                             39              39                     39   
IN                             30              30                     30   
KS                             15              15                     15   
KY                             17              17                     17   
MD                             22              22                     22   
ME                             11              11                     11   
MI                             37              37                     37   
MO                             16              16                     16   
MS                             10              10                     10   
MT                              6               6                      6   
NC                             20              20                     20   
ND                              1               1                      1   
NE                             12              12                     12   
NJ                             15              15                     15   
NM                             11              11                     11   


As you can see in my data above, this data set is pretty stable and clean, as the number is constant throughout all of the columns.

**Sorting Candidates by District:**

In [155]:
#I can do the same thing for district and gather more information on 
#how many candidates ran for each district, and I am able to display those unique districts!

df3.groupby("district").count()

candidate  state  office_type  race_type  \
district                                                                        
Governor of Alabama                          6      6            6          6   
Governor of Arkansas                         2      2            2          2   
Governor of California                      12     12           12         12   
Governor of Colorado                         4      4            4          4   
Governor of Georgia                          2      2            2          2   
Governor of Idaho                            3      3            3          3   
Governor of Illinois                         6      6            6          6   
Governor of Iowa                             6      6            6          6   
Governor of Kansas                           5      5            5          5   
Governor of Maine                            7      7            7          7   
Governor of Maryland                         8      8            8          8   
Governor of Michigan                         3      3            3          3   
Governor of Nebraska                         3      3            3          3   
Governor of Nevada                           6      6            6          6   
Governor of New Mexico                       3      3            3          3   
Governor of Ohio                             6      6            6          6   
Governor of Oklahoma                         2      2            2          2   
Governor of South Carolina                   3      3            3          3   
Governor of South Dakota                     1      1            1          1   
Governor of Tennessee                        3      3            3          3   
Governor of Texas                            9      9            9          9   
U.S. House Alabama District 1                2      2            2          2   
U.S. House Alabama District 2                2      2            2          2   
U.S. House Alabama District 3                2      2            2          2   
U.S. House Alabama District 4                2      2            2          2   
U.S. House Alabama District 5                1      1            1          1   
U.S. House Alabama District 6                1      1            1          1   
U.S. House Arizona District 8                2      2            2          2   
U.S. House Arkansas District 1               1      1            1          1   
U.S. House Arkansas District 2               4      4            4          4   
...                                        ...    ...          ...        ...   
U.S. House Texas District 4                  2      2            2          2   
U.S. House Texas District 5                  1      1            1          1   
U.S. House Texas District 6                  5      5            5          5   
U.S. House Texas District 7                  7      7            7          7   
U.S. House Texas District 8                  1      1            1          1   
U.S. House Utah District 1                   2      2            2          2   
U.S. House Utah District 2                   1      1            1          1   
U.S. House Utah District 3                   1      1            1          1   
U.S. House Utah District 4                   1      1            1          1   
U.S. House Virginia District 1               3      3            3          3   
U.S. House Virginia District 10              6      6            6          6   
U.S. House Virginia District 2               2      2            2          2   
U.S. House Virginia District 5               1      1            1          1   
U.S. House Virginia District 6               4      4            4          4   
U.S. House Virginia District 7               2      2            2          2   
U.S. House Virginia District 9               2      2            2          2   
U.S. House Washington District 3             4      4            4  

_ **Areas of Future Exploration:** _

**Counting LGBTQ Candidates (Incomplete, read below):**

In [156]:
#for some reason, I did not get a result from counting the groupby of "lgbtq", so I called the specific index instead 
#the same thing happened for race
#maybe it's because these columns are not visibile with the head function? Or maybe I need to explore more about type 
#conversion in the future 

df3.groupby(list(df3)[14]).count
df3.groupby('race').count

<bound method DataFrameGroupBy.count of <pandas.core.groupby.groupby.DataFrameGroupBy object at 0x1a10809160>>

For Future Exploration for Above:
- Why couldn't it count the column "lgbtq" or "race"? 
- Explore ways to preform type conversion 

**Trying out describe for the primary runoff info (more future exploration):**

In [157]:
#df3('primary_runoff_status').describe()

#this results in an error as the 'dataframe' object isn't callable. I think it is because maybe this column was read 
#in as a string
#This is another item of future exploration, as I will have to figure out how to turn this column into an int or float 
#in order for the function describe to work. This would be my next step. 

**Data Analysis and Conclusion**

Original Analysis:

>Overall, I learned alot from the data set and analysis at hand. Out of the 811 democratic candidates that this data set covered, around 12% of them ran for governor, while around 84% ran for U.S. House of Representatives. 4.6% of candidates are LGBTQ, and 25% of them do not identify as white. About 15% of candidates who ran are veterans. The majority of candidates lost their race, but the winning democrat moves onto the general election- as the 265 advancing are the democratic candidate in the general. A large amount of democratic candidates ran for office this year, as more candidates lost than one (as they were racing against eachother). This should mean a good turnout of voters and support for the general, as the democratic winner was narrowed against a tougher competition with more candidates. This large show of democratic candidates suggests a strong support for democrats in November, but could also lead to a split vote as more candidates ran and a greater variety were supported. In addition, this year a fair amount of LGBTQ, non-white, and veterans ran for office, making for a more diverse and different group of candidates. 

Updated Analysis:

Through my new knowledge of data cleaning and pandas, I was able to gather additional information surrounding my data set. Out of the 811 democratic candidates, 100 ran for governor, 687 ran for U.S. House of Reps, and 24 ran for senator positions. 265 candidates advanced, while 546 did not. My data set covers primaries from a total of 37 states. Of these 37 states, the most candidates ran in Texas, as they has a total of 109 candidates who ran. The second highest was California, with 68 candidates. Out of the entire data set, the district in which the most candidates ran in the country was for governor of california, with 12 candidates for this position.

Overall, through pandas, I was able to gather more specific data in a wider range. In the future, I would love to explore methods beyond count to work with when using string values. If I had more time, I would also have loved to find a way to convert my strings to integers per column and then be able to analyze that with pandas methods such as describe and info. Again, the large number of democratic candidates who ran should mean a good turnout of voters as the primary elections were more competitive. However, again, this could also lead to a split vote as a greater pool of candidates ran in the primary. Additionally, I was surprised to find that there were 24 democratic candidates who ran for senator seats. Those senate races will be interesting ones to watch. Because there is a smaller pool of candidates for these races, and senate seats are more competitive, the outcomes of the senate races could be super important. Finally, the governor of California was the most competitive primary democratic race included in this set of data, with 12 candidates running for one position. This should be interesting to watch. Because there were 12 candidates to choose from, this may mean more democratic support in the general as their candidates all had tough competition. However, this could again just mean a more split vote as less people's vote were represented over the 12 candidates. 

**Acknowledgements**

For this project, I referred to the More Python Practice file that Ms. Sconyers shared with us to recall certain functions for calling, reading, and organizing data. I worked with Ms. Sconyers alot on how to organize my data. Additionally, I used John Gruber's Canonical Markdown Cheatsheet. All data was pulled from Five Thirty Eight.

When reviewing this lab, I used the pandas cheat-cheat a lot as well. (https://www.dataquest.io/blog/large_files/pandas-cheat-sheet.pdf) 

I also greatly relied on my Lab 4 during the process both of cleaning and utilizing pandas in my data set, where I gathered information from a variety of online sources, Ms. Sconyers, and Julia.  